### ISSUE TITLE
knex no longer spyable by jest ?

### ISSUE THREAD

**Issue by mlarcher** [knex/knex#3775]
# Environment

Knex version: 0.20.11
OS: macos

# Bug

When updating from 0.20.10 to 0.20.11, test suite breaks because Jest spies cannot be applied.
For example, `jest.spyOn(db, 'destroy')` or `jest.spyOn(db, 'transaction')` throw the following error :
```js
TypeError: Cannot assign to read only property 'transaction' of function 'function knex(tableName, options) {
        return createQueryBuilder(knex.context, tableName, options);
      }'
```
Is this intentional ? Is there a new way of doing things to get around this problem ?
I'd be happy o provide additional information if needed

**Comment by briandamaged** [comment]
Howdy!  We cleaned up some of the underlying code and fixed the way it was proxying some function calls.  So, you might need to spy on `knex.context` instead of just `knex` .

**Comment by mlarcher** [comment]
So the spy needs to be on something else than what we use? Or should we go through context when initiating transactions too ?

**Comment by briandamaged** [comment]
The spy itself will need to be applied to `db.context` instead of just `db`.  

It's still safe to call `db.transact(..)`.  This method is ultimately a shortcut for `db.context.transact(..)` now.

And yeah: sorry about making `db.transact` property read-only for now.  There were a few places in the `knex` code that were mistakenly injecting behaviors into `db` instead of `db.context`.  So, we marked these proxy methods as read-only to help us catch these mistakes.  (Basically: these mistakes could lead to some really baffling bugs that were difficult to track down)

**Comment by mlarcher** [comment]
Thank you for third insight !
Are you saying that this shouldn't last long and readonly should be removed once you think you have caught all unwanted injections? If so, we might stay with our version until it is the case... Or please tell us if that may last long for some reason...
I think having the spy on a different path thank the one used leads to a bad developer experience, so we have to weight our options here

**Comment by mlarcher** [comment]
I tried to update the spies so that they point to db.context methods, things went fine for those but something else failed : 
```js
TypeError: uuid.v1 is not a function

      126 |     .where('talent.id', talentId)
      127 |     .leftJoin('user', 'user.id', 'talent.userId')
    > 128 |     return a.then((rows) => {
          |              ^
      129 |       if (!rows[0]) {
      130 |         return null;
      131 |       }

      at QueryCompiler_PG.toSQL (node_modules/knex/lib/query/compiler.js:72:28)
      at Builder.toSQL (node_modules/knex/lib/query/builder.js:77:44)
      at node_modules/knex/lib/runner.js:31:36
      at node_modules/knex/lib/runner.js:260:24
      From previous event:
      From previous event:
      at Runner.ensureConnection (node_modules/knex/lib/runner.js:258:8)
      at Runner.run (node_modules/knex/lib/runner.js:26:12)
      at Builder.then (node_modules/knex/lib/interface.js:16:43)
      at Object.findByTalentId (src/domains/user/userService.js:128:14)
      at setup (src/domains/user/__tests__/userService.test.js:269:26)
      at Object.<anonymous> (src/domains/user/__tests__/userService.test.js:284:26)
```

**Comment by briandamaged** [comment]
> Are you saying that this shouldn't last long and readonly should be removed once you think you have caught all unwanted injections?

Hmm... that's a good question.  For now, it's probably best to assume that they will remain read-only.

The reason: basically, the `knex` function is just a convenient proxy for the `context` and `client` objects.  However, due to some quirks in how the proxy behavior was originally implemented, it allowed for a lot of weird "blurring" to occur across these objects.  For example: when you invoked `knex.transaction(..)`, it would:

1. Call, `knex.context.transaction(..)`, but
2. set `this` to `knex`!  (Instead of `knex.context`, like you would expect)

And, due to this weird object blurring, other bugs were becoming obfuscated as well.  For instance: sections of the code would override `knex.transaction(..)`, which would magically work thanks to the object blurring bug.  In other words: one bug was cancelling the other bug.

So yeah: for now, it would probably be best for the test cases to spy on `knex.context` instead of `knex` itself.  Otherwise, your test cases might start failing due to bugs that are no longer being canceled out.

> I tried to update the spies so that they point to db.context methods, things went fine for those but something else failed

Hmm... that looks like an unrelated issue.  Which version of the `uuid` library are you currently using?  (Or, is it possible that you stubbed out something in the `uuid` library in one of your other tests?)

**Comment by mlarcher** [comment]
Thanks again for all this information :)
About uuid, you are correct : it changed recently and we had to mock the whole module instead of just spying on the target method, so it was impacting knex too. I exposed a .v1() stub method in the tests for use by knex and all tests are green again now.

Thank you for your help. I'll leave it to you to close or not the issue

**Comment by briandamaged** [comment]
@mlarcher : np!

BTW: there seems to be an open discussion in #3774 about the impact of the `uuid.v1()` change.  You might want to keep an eye on that thread just in case any decisions made there would impact your codebase.

**Comment by kibertoad** [comment]
Fix for the uuid change was released in 0.21.10

**Comment by mlarcher** [comment]
thank you for the update :)
